python/GafferArnoldTest/ArnoldRenderTest.py

```diff
@@ -1478,5 +1478,71 @@ def testCoordinateSystem( self ) :
 			# node to have been created for ours.
 			self.assertIsNone( arnold.AiNodeLookUpByName( universe, "/coordinateSystem" ) )
 
+	@GafferTest.TestRunner.PerformanceTestMethod( repeat = 1 )
+	def testInstancerPerf( self ) :
```

Comment:
Concerningly, my numbers look completely different:

```
With InstancerCapsule
=====================

1 thread
--------

testInstancerEncapsulatePerf : 23.3s
testInstancerPerf            : 10.4s

All threads
-----------

testInstancerEncapsulatePerf : 3.27s
testInstancerPerf            : 1.65s

Without InstancerCapsule
========================

1 thread
--------

testInstancerEncapsulatePerf : 148s
testInstancerPerf : 10.5s

All threads
-----------

testInstancerEncapsulatePerf  : 19.4s
testInstancerPerf : 1.57s
```

I'm not seeing noticeable improvement in `testInstancerPerf`, but I am seeing 5X improvements in `testInstancerEncapsulatePerf` in both single threaded and multithreaded. Obviously, getting an improvement in `testInstancerPerf` is not a concern - I did try to optimize what I could while doing the refactor to get rid of prototypeChildNames ( maybe now that we're ready for review, it would be pretty helpful if I try and split out that change separately - in practice it all happened together, but presenting it separately might be clearer ).

But it's a pretty dramatic concern if you're not seeing the same improvement I am when using multithreading ... I'm worried about how I'm going to analyze this if it happens differently on your system. What results do you get with -threads 16?

Reply:
For reference, after switching testInstancerEncapsulatePerf to 500 divisions to match testInstancerPerf, here are my new numbers. Still seeing consistent 5-6X improvements with encapsulating.

The initial advantages of encapsulating without InstancerCapsule are around 15%, which seems inline with my observations yesterday about AiUniverseDestroy being much slower for things not wrapped in a procedural.

```
With InstancerCapsule
=====================

1 thread
--------

testInstancerEncapsulatePerf : 1.46s 
testInstancerPerf : 10.9s 


All threads
-----------

testInstancerEncapsulatePerf : 0.211s
testInstancerPerf : 1.64s 

Without InstancerCapsule
========================

1 thread
--------

testInstancerEncapsulatePerf : 9.25s
testInstancerPerf : 10.4s

All threads
-----------

testInstancerEncapsulatePerf : 1.23s
testInstancerPerf : 1.48s
```

Reply:
Here are some more results with your latest push, for a variety of threadcounts :

```
                                  With InstancerCapsule         Without InstancerCapsule
=======================================================================================

64 threads
----------

testInstancerEncapsulatePerf :   0.723s                        0.953s
testInstancerPerf :              2.12s                         1.81s

32 threads
----------

testInstancerEncapsulatePerf :   0.374s                        1.06s
testInstancerPerf :              1.45s                         1.33s

16 threads
----------

testInstancerEncapsulatePerf :   0.449s                        1.3s
testInstancerPerf :              1.68s                         1.59s

8 threads
---------

testInstancerEncapsulatePerf :   0.544s                        2s
testInstancerPerf :              2.4s                          2.46s

4 threads
---------

testInstancerEncapsulatePerf :   0.778s                        3.3
testInstancerPerf :              4.06s                         3.96

2 threads
---------

testInstancerEncapsulatePerf :   1.04s                         5.43s
testInstancerPerf :              6.83s                         7.34s

1 threads
---------

testInstancerEncapsulatePerf :   1.53s
testInstancerPerf :              9.26s
```

So I do see a more pronounced improvement up to 32 threads, but then 64 is slower than 32 for the new code. I have 32 physical cores and those results are with hyperthreading on. Turning hyperthreading off I get an improvement up to 16 cores, and then performance gets worse at 32. I don't know what to make of that!